activesupport/lib/active_support/lazy_load_hooks.rb

```diff
@@ -40,20 +46,32 @@ def on_load(name, options = {}, &block)
       @load_hooks[name] << [block, options]
     end
 
-    def execute_hook(base, options, block)
-      if options[:yield]
-        block.call(base)
-      else
-        base.instance_eval(&block)
-      end
-    end
-
     def run_load_hooks(name, base = Object)
       @loaded[name] << base
       @load_hooks[name].each do |hook, options|
         execute_hook(base, options, hook)
       end
     end
+
+    private
+
+      def with_execution_control(name, block, once)
+        unless @run_once[name].include?(block)
+          @run_once[name] << block if once
+
+          yield
+        end
+      end
+
+      def execute_hook(base, options, block)
+        with_execution_control(name, block, options[:run_once]) do
```

Comment:
Where is `name` coming from here? Previously `execute_hook` was changed to pass it in.